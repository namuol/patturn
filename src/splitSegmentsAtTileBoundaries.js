// @flow
import type {Path} from './types';

type Params = {
  path: Path,
  tileWidth: number,
  tileHeight: number,
};

const crossesGrid = (v1, v2, size) => {
  const cell1 = Math.floor(v1 / size);
  const cell2 = Math.floor(v2 / size);
  return cell1 !== cell2;
};

const splitPathAtTileBoundaries = (
  {path: {points, strokeWidth, smoothFactor}, tileWidth, tileHeight}: Params,
): Array<Path> => {
  let lastPoint = points[0];
  const paths = [{points: [lastPoint], strokeWidth, smoothFactor}];
  for (let index = 1; index < points.length; index += 1) {
    const point = points[index];
    const {x: x1, y: y1} = lastPoint;
    const {x: x2, y: y2} = point;

    if (crossesGrid(x1, x2, tileWidth) || crossesGrid(y1, y2, tileHeight)) {
      paths.push(
        {
          points: [lastPoint, point],
          intersectsGrid: true,
          strokeWidth,
          smoothFactor,
        },
        {points: [point], strokeWidth, smoothFactor},
      );
    } else {
      paths[paths.length - 1].points.push(point);
    }

    lastPoint = point;
  }

  return paths;
};

export default splitPathAtTileBoundaries;
