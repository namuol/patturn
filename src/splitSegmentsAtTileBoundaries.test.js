// @flow
import splitSegmentsAtTileBoundaries from './splitSegmentsAtTileBoundaries';

it('returns a _set_ of paths', () => {
  // prettier-ignore
  const points = [
    {x: 10, y: 10},
    {x: 20, y: 20},
  ];

  expect(
    splitSegmentsAtTileBoundaries({
      path: {points},
      tileWidth: 100,
      tileHeight: 100,
    }),
  ).toEqual([{points}]);
});

it('splits the path into 3 when crossing from the left', () => {
  // prettier-ignore
  const points = [
    {x: 20, y: 0},
    {x: 10, y: 0},
    {x: -10, y: 0},
    {x: -20, y: 0},
  ];

  expect(
    splitSegmentsAtTileBoundaries({
      path: {points},
      tileWidth: 100,
      tileHeight: 100,
    }),
  ).toMatchSnapshot();
});

it('splits the path into 3 when crossing from the right', () => {
  // prettier-ignore
  const points = [
    {x: -20, y: 0},
    {x: -10, y: 0},
    {x: 10, y: 0},
    {x: 20, y: 0},
  ];

  expect(
    splitSegmentsAtTileBoundaries({
      path: {points},
      tileWidth: 100,
      tileHeight: 100,
    }),
  ).toMatchSnapshot();
});

it('splits the path into 3 when crossing from the left far inside the grid', () => {
  // prettier-ignore
  const points = [
    {x: 500 + 20, y: 0},
    {x: 500 + 10, y: 0},
    {x: 500 + -10, y: 0},
    {x: 500 + -20, y: 0},
  ];

  expect(
    splitSegmentsAtTileBoundaries({
      path: {points},
      tileWidth: 100,
      tileHeight: 100,
    }),
  ).toMatchSnapshot();
});

it('splits the path into 3 when crossing from the right far inside the grid', () => {
  // prettier-ignore
  const points = [
    {x: 500 + -20, y: 0},
    {x: 500 + -10, y: 0},
    {x: 500 + 10, y: 0},
    {x: 500 + 20, y: 0},
  ];

  expect(
    splitSegmentsAtTileBoundaries({
      path: {points},
      tileWidth: 100,
      tileHeight: 100,
    }),
  ).toMatchSnapshot();
});

// Vertical

it('splits the path into 3 when crossing from the bottom', () => {
  // prettier-ignore
  const points = [
    {x: 0, y: 20},
    {x: 0, y: 10},
    {x: 0, y: -10},
    {x: 0, y: -20},
  ];

  expect(
    splitSegmentsAtTileBoundaries({
      path: {points},
      tileWidth: 100,
      tileHeight: 100,
    }),
  ).toMatchSnapshot();
});

it('splits the path into 3 when crossing from the top', () => {
  // prettier-ignore
  const points = [
    {x: 0, y: -20},
    {x: 0, y: -10},
    {x: 0, y: 10},
    {x: 0, y: 20},
  ];

  expect(
    splitSegmentsAtTileBoundaries({
      path: {points},
      tileWidth: 100,
      tileHeight: 100,
    }),
  ).toMatchSnapshot();
});

it('splits the path into 3 when crossing from the bottom far inside the grid', () => {
  // prettier-ignore
  const points = [
    {x: 0, y: 500 + 20},
    {x: 0, y: 500 + 10},
    {x: 0, y: 500 + -10},
    {x: 0, y: 500 + -20},
  ];

  expect(
    splitSegmentsAtTileBoundaries({
      path: {points},
      tileWidth: 100,
      tileHeight: 100,
    }),
  ).toMatchSnapshot();
});

it('splits the path into 3 when crossing from the top far inside the grid', () => {
  // prettier-ignore
  const points = [
    {x: 0, y: 500 + -20},
    {x: 0, y: 500 + -10},
    {x: 0, y: 500 + 10},
    {x: 0, y: 500 + 20},
  ];

  expect(
    splitSegmentsAtTileBoundaries({
      path: {points},
      tileWidth: 100,
      tileHeight: 100,
    }),
  ).toMatchSnapshot();
});
