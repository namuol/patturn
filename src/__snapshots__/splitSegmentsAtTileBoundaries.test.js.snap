exports[`test splits the path into 3 when crossing from the bottom 1`] = `
Array [
  Object {
    "points": Array [
      Array [
        0,
        20,
      ],
      Array [
        0,
        10,
      ],
    ],
  },
  Object {
    "intersectsGrid": true,
    "points": Array [
      Array [
        0,
        10,
      ],
      Array [
        0,
        -10,
      ],
    ],
  },
  Object {
    "points": Array [
      Array [
        0,
        -10,
      ],
      Array [
        0,
        -20,
      ],
    ],
  },
]
`;

exports[`test splits the path into 3 when crossing from the bottom far inside the grid 1`] = `
Array [
  Object {
    "points": Array [
      Array [
        0,
        520,
      ],
      Array [
        0,
        510,
      ],
    ],
  },
  Object {
    "intersectsGrid": true,
    "points": Array [
      Array [
        0,
        510,
      ],
      Array [
        0,
        490,
      ],
    ],
  },
  Object {
    "points": Array [
      Array [
        0,
        490,
      ],
      Array [
        0,
        480,
      ],
    ],
  },
]
`;

exports[`test splits the path into 3 when crossing from the left 1`] = `
Array [
  Object {
    "points": Array [
      Array [
        20,
        0,
      ],
      Array [
        10,
        0,
      ],
    ],
  },
  Object {
    "intersectsGrid": true,
    "points": Array [
      Array [
        10,
        0,
      ],
      Array [
        -10,
        0,
      ],
    ],
  },
  Object {
    "points": Array [
      Array [
        -10,
        0,
      ],
      Array [
        -20,
        0,
      ],
    ],
  },
]
`;

exports[`test splits the path into 3 when crossing from the left far inside the grid 1`] = `
Array [
  Object {
    "points": Array [
      Array [
        520,
        0,
      ],
      Array [
        510,
        0,
      ],
    ],
  },
  Object {
    "intersectsGrid": true,
    "points": Array [
      Array [
        510,
        0,
      ],
      Array [
        490,
        0,
      ],
    ],
  },
  Object {
    "points": Array [
      Array [
        490,
        0,
      ],
      Array [
        480,
        0,
      ],
    ],
  },
]
`;

exports[`test splits the path into 3 when crossing from the right 1`] = `
Array [
  Object {
    "points": Array [
      Array [
        -20,
        0,
      ],
      Array [
        -10,
        0,
      ],
    ],
  },
  Object {
    "intersectsGrid": true,
    "points": Array [
      Array [
        -10,
        0,
      ],
      Array [
        10,
        0,
      ],
    ],
  },
  Object {
    "points": Array [
      Array [
        10,
        0,
      ],
      Array [
        20,
        0,
      ],
    ],
  },
]
`;

exports[`test splits the path into 3 when crossing from the right far inside the grid 1`] = `
Array [
  Object {
    "points": Array [
      Array [
        480,
        0,
      ],
      Array [
        490,
        0,
      ],
    ],
  },
  Object {
    "intersectsGrid": true,
    "points": Array [
      Array [
        490,
        0,
      ],
      Array [
        510,
        0,
      ],
    ],
  },
  Object {
    "points": Array [
      Array [
        510,
        0,
      ],
      Array [
        520,
        0,
      ],
    ],
  },
]
`;

exports[`test splits the path into 3 when crossing from the top 1`] = `
Array [
  Object {
    "points": Array [
      Array [
        0,
        -20,
      ],
      Array [
        0,
        -10,
      ],
    ],
  },
  Object {
    "intersectsGrid": true,
    "points": Array [
      Array [
        0,
        -10,
      ],
      Array [
        0,
        10,
      ],
    ],
  },
  Object {
    "points": Array [
      Array [
        0,
        10,
      ],
      Array [
        0,
        20,
      ],
    ],
  },
]
`;

exports[`test splits the path into 3 when crossing from the top far inside the grid 1`] = `
Array [
  Object {
    "points": Array [
      Array [
        0,
        480,
      ],
      Array [
        0,
        490,
      ],
    ],
  },
  Object {
    "intersectsGrid": true,
    "points": Array [
      Array [
        0,
        490,
      ],
      Array [
        0,
        510,
      ],
    ],
  },
  Object {
    "points": Array [
      Array [
        0,
        510,
      ],
      Array [
        0,
        520,
      ],
    ],
  },
]
`;
